### PR TITLE
Initializes PartiQL's Code Coverage library

### PR DESCRIPTION
**Pull request by johnedquinn**
## Relevant Issues
- None

## General Description
- Initializes PartiQL's Code Coverage library
  - Adds support for BRANCH Coverage
  - Adds support for BRANCH-CONDITION Coverage
  - Adds integration with JUnit5 for ease-of-use
  - Adds documentation to GH Wiki

## Description
- Adds the ability to gather code coverage statistics for compiled/executed PartiQL statements. Accomplishes this by:
  - Adding static `CoverageStructure` to the compiled `Expression`
  - Adding `CoverageData` to the evaluated `PartiQLResult`
  - Wrapping the `EvaluatingCompiler` with a `CoverageCompiler`. This essentially gathers information for the `CoverageStructure` and `CoverageData`
- Initializes a new project called `partiql-coverage` to integrate with JUnit and allow for users to test their PartiQL statements. This is accomplished by:
  - Exposing a `PartiQLTest` annotation that implements JUnit's `@Test`
  - The `PartiQLTest` annotation invokes the `PartiQLTestExtension` which receives the `PartiQLTestProvider`.
  - The `PartiQLTestExtension` compiles the PartiQL statement (and grabs the `CoverageStructure`) found in the `PartiQLTestProvider`, then, it loops through all test cases provided by the `PartiQLTestProvider`. It passes the test case and the result to the invoking test method (so that users can perform assertions). It then aggregates the `CoverageData` from each result and reports relevant coverage statistics to registered JUnit listeners.
- Updates Gradle from 7.5.1 to 7.6.2. Not breaking. Did this so I can use `--rerun` which is quite convenient.

## Other Information
- Updated Unreleased Section in CHANGELOG: **YES**
- Any backward-incompatible changes? **NO**
  - No.
- Any new external dependencies? **YES**
  - Not to any existing libraries. However, the new package `partiql-coverage` depends on JUnit as well as JGenHtml.
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-coverage/README.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+1. While PartiQL’s current evaluator does not statically resolve the types of expressions, it is unknown whether a
+   specific branch-condition might return TRUE, FALSE, NULL, or MISSING. Therefore, the existing Evaluating Compiler will,
+   based on the typing mode, assume that all boolean expressions can return ALL possible outcomes. In the future, when
+   constant-folding and static-typing are part of the planning process, this will be mitigated.
+2. Similar to #1, as the current evaluate does not statically resolve the types of expressions, functions and variable
+   references will not be part of a Branch-Condition Coverage result. However, in the future, support for this will be added.
+3. As there is not a consensus regarding how `COALESCE` and `NULLIF` will be represented in the future evaluator,
+   these two SQL-defined functions currently do not create branches. However, this is subject to change in future releases.
+
```
Comment (am357) [inline partiql-coverage/README.md:89]: Just curious: does this mean that `COALESCE`, `NULLIF`, `WHERE`, `HAVING`, and `CASE_WHEN` are all the control flows in the scope or there are others too? maybe worthwhile having a pointer to this in the Wiki.

File: test/coverage-tests/src/test/kotlin/org/partiql/test/coverage/RunnerTest.kt [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    @PartiQLTest(provider = InitializationFailureProvider::class)
+    fun initializationFailureExample(tc: InitializationFailureProvider.TestCase, result: ExprValue) {
+        assertEquals(ExprValueType.BOOL, result.type)
+    }
+
+    @Ignore
+    @PartiQLTest(provider = TestInitializationFailureProvider::class)
+    fun testInitializationFailureExample(tc: TestInitializationFailureProvider.TestCase, result: ExprValue) {
+        assertEquals(ExprValueType.STRING, result.type)
+    }
```
Comment (am357) [inline test/coverage-tests/src/test/kotlin/org/partiql/test/coverage/RunnerTest.kt:60]: Just double-checking: do we want to ignore these?

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/CoverageCompiler.kt [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+     // conditionCounts.replaceAll { _, _ -> 0 }
+     // branchCounts.replaceAll { _, _ -> 0 }
+     CoverageData(branchConditionCount = returnedConditionCounts, branchCount = returnedBranchCounts)
+     },
+     coverageStructure = { coverageStructure }
+     )
+     }
+     }
+     }
+     **/
```
Comment (am357) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/CoverageCompiler.kt:141]: Intentional?

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/CoverageCompiler.kt [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+    // TODO: Figure out how we can determine whether an ID should be a boolean.
```
Comment (am357) [inline partiql-lang/src/main/kotlin/org/partiql/lang/eval/impl/CoverageCompiler.kt:143]: Could you please elaborate this and the next comment? I guess I'm missing a prior context.

**Review by am357 — COMMENTED** [review]
Excellent addition to the library—thank you for all the effort!

LGTM—left some non-blocking comments. There is only one related to a code block comment which perhaps makes sense to remove.

File: test/coverage-tests/src/test/kotlin/org/partiql/test/coverage/RunnerTest.kt (hunk 2 above)
Comment (johnedquinn) [reply test/coverage-tests/src/test/kotlin/org/partiql/test/coverage/RunnerTest.kt]: Yes, it was to just show that we can use those annotations. Adding a comment to clarify, thanks!

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/Environment.kt [hunk 5]
```diff
@@ -30,7 +30,9 @@ internal data class Environment(
     val current: Bindings<ExprValue> = locals,
     val session: EvaluationSession,
     val groups: MutableMap<ExprValue, Group> = createGroupMap(),
-    val currentGroup: Group? = null
+    val currentGroup: Group? = null,
+    val branchCounts: MutableMap<String, Long>? = null,
+    val branchConditionCounts: MutableMap<String, Long>? = null
```
[the remark this answers is not in the document]
Comment (am357) [reply partiql-lang/src/main/kotlin/org/partiql/lang/eval/Environment.kt:35]: Sounds good

File: partiql-coverage/README.md (hunk 1 above)
Comment (am357) [reply partiql-coverage/README.md:89]: Line 29 conflates what we support and don't. Perhaps would make sense to make what we support and don't out of all more explicit.

**Review by am357 — APPROVED** [review]

**Pull request merged**